src/Microsoft.DotNet.ProjectJsonMigration/ProjectMigrator.cs

```diff
@@ -33,14 +35,51 @@ public ProjectMigrator(IMigrationRule ruleSet)
             _ruleSet = ruleSet;
         }
 
-        public void Migrate(MigrationSettings migrationSettings)
+        public void Migrate(MigrationSettings settings, bool skipProjectReferences = false)
         {
-            var migrationRuleInputs = ComputeMigrationRuleInputs(migrationSettings);
-            VerifyInputs(migrationRuleInputs, migrationSettings);
+            if (settings == null)
+            {
+                throw new ArgumentNullException();
+            }
 
-            SetupOutputDirectory(migrationSettings.ProjectDirectory, migrationSettings.OutputDirectory);
+            Queue<MigrationSettings> settingsQueue = new Queue<MigrationSettings>();
+            settingsQueue.Enqueue(settings);
+
+            while(settingsQueue.Count() != 0)
+            {
+                var migrationSettings = settingsQueue.Dequeue();
+                var migrationRuleInputs = ComputeMigrationRuleInputs(migrationSettings);
+                VerifyInputs(migrationRuleInputs, migrationSettings);
 
-            _ruleSet.Apply(migrationSettings, migrationRuleInputs);
+                SetupOutputDirectory(migrationSettings.ProjectDirectory, migrationSettings.OutputDirectory);
+
+                _ruleSet.Apply(migrationSettings, migrationRuleInputs);
+                _migratedProjects.Add(migrationSettings.ProjectDirectory);
+
+                if (skipProjectReferences)
+                {
+                    continue;
+                }
+
+                var projectContext = migrationRuleInputs.DefaultProjectContext;
+                var projectExports = projectContext.CreateExporter("_").GetDependencies(LibraryType.Project);
+
+                foreach(var export in projectExports)
+                {
+                    var projectFile = ((ProjectDescription)export.Library).Project.ProjectFilePath;
+                    var projectDir = Path.GetDirectoryName(projectFile);
+
+                    if (_migratedProjects.Contains(projectDir))
```

Comment:
You can have a case where a project is added to the queue multiple times before it's added to _migratedProjects causing duplicate work. 

One solution would be to check _migratedProjects immediately after Dequeueing, and just let the duplicates get on the queue but then ignore them.